pkg/filesystem/path/parser.go

```diff
@@ -0,0 +1,150 @@
+			return p
+		}
+	}
+}
+
+func (p *unixAbsoluteParser) ParseScope(scopeWalker ScopeWalker) (next ComponentWalker, remainder RelativeParser, err error) {
+	absolute := false
+	if p.path != "" && p.path[0] == '/' {
+		p.path = stripOneOrMoreSlashes(p.path)
+		absolute = true
+	}
+
+	if absolute {
+		next, err = scopeWalker.OnAbsolute()
+	} else {
+		next, err = scopeWalker.OnRelative()
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+
+	remainder, err = newRelativePath(p.path)
+	if err != nil {
+		return nil, nil, err
+	}
+	return next, remainder, nil
+}
+
+func (rp unixRelativeParser) ParseFirstComponent(componentWalker ComponentWalker, mustBeDirectory bool) (next GotDirectoryOrSymlink, remainder RelativeParser, err error) {
+	var name string
+	terminal := false
+	if slash := strings.IndexByte(rp.path, '/'); slash == -1 {
+		// Path no longer contains a slash. Consume it entirely.
+		terminal = true
+		name = rp.path
+		remainder = nil
+	} else {
+		name = rp.path[:slash]
+		rp.path = stripOneOrMoreSlashes(rp.path[slash:])
+		remainder = unixRelativeParser{rp.path}
+	}
+
+	switch name {
+	case "", ".":
+		// An explicit "." entry, or an empty component.
+		// Empty components can occur if paths end with
+		// one or more slashes. Treat "foo/" as identical
+		// to "foo/."
+		return GotDirectory{Child: componentWalker}, remainder, nil
+	case "..":
+		// Traverse to the parent directory.
+		parent, err := componentWalker.OnUp()
+		if err != nil {
+			return nil, nil, err
+		}
+		return GotDirectory{Child: parent}, remainder, nil
+	}
+
+	// A filename that was followed by a
+	// slash, or we are symlink expanding
+	// one or more paths that are followed
+	// by a slash. This component must yield
+	// a directory or symlink.
+	if mustBeDirectory || !terminal {
+		r, err := componentWalker.OnDirectory(Component{
+			name: name,
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+		next = r
+		return next, remainder, nil
+	}
+
+	r, err := componentWalker.OnTerminal(Component{
+		name: name,
+	})
+	if err != nil || r == nil {
+		// Path resolution ended with
+		// any file other than a symlink.
+		return nil, nil, err
+	}
+
+	// Observed a symlink at the end of a
+	// path. We should continue to run.
+	return GotSymlink{
+		Parent: r.Parent,
+		Target: r.Target,
+	}, remainder, nil
+}
+
+// NewUNIXParser creates a Parser for Unix paths that can be used in Resolve.
+func NewUNIXParser(path string) (Parser, error) {
+	// Unix-style paths are generally passed to system calls that
+	// accept C strings. There is no way these can accept null
+	// bytes.
+	if strings.ContainsRune(path, '\x00') {
+		return nil, status.Error(codes.InvalidArgument, "Path contains a null byte")
+	}
+
+	return &unixAbsoluteParser{path}, nil
+}
+
+func newRelativePath(path string) (unixRelativeParser, error) {
+	if len(path) > 0 && path[0] == '/' {
+		return unixRelativeParser{}, status.Errorf(codes.InvalidArgument, "Invalid relative path %#v", path)
+	}
```

Comment:
I'm not sure this error can actually occur. I mean, we only construct relative paths within this specific source file, and there we always make sure to strip slashes.

Reply:
Good point! Replaced with a bare construction.